### ISSUE TITLE
It's not possible to edit "legacy" DOS files ending in ^Z

### ISSUE THREAD

**Issue by arp242** [vim/vim#11397]
Old systems such as CP/M, MSX, and some versions of DOS used CRLF line endings and a ^Z at the end of the file; for example:

    % cat -v msx.bas
    10 ? "HELLO"^M
    20 GOTO 10^M
    ^Z

    [~]% hexdump -C msx.bas
    00000000  31 30 20 3f 20 22 48 45  4c 4c 4f 22 0d 0a 32 30  |10 ? "HELLO"..20|
    00000010  20 47 4f 54 4f 20 31 30  0d 0a 1a                 | GOTO 10...|

Currently, when fileformat=dos is used it will always drop the ^Z at the end, as documented:

> When "dos" is used, CTRL-Z at the end of a file is ignored. See file-formats and file-read.

This is disabled when 'binary' is set, but then it always uses Unix line endings, for example after editing that file with `vim --clean -b msx.bas` and adding a line I end up with:

    00000000  31 30 20 3f 20 22 48 45  4c 4c 4f 22 0d 0a 31 35  |10 ? "HELLO"..15|
    00000010  20 3f 20 22 57 4f 52 4c  44 22 0a 32 30 20 47 4f  | ? "WORLD".20 GO|
    00000020  54 4f 20 31 30 0d 0a 1a                           |TO 10...|

The second line has a LF, rather than CRLF, but the previous lines and ^Z is preserved.

---

I'm not sure what the best fix for this would be? I know these are really old files for long-since obsolete systems, but there are still enthusiasts playing around with these systems. 

Adding an entire new option for this seems a bit meh, but could be done I suppose. Maybe a way to make 'binary' use CRLF line endings? Although that also has other side-effects, it may be enough for this relatively obscure use case(?)

**Comment by k-takata** [comment]
If you append ^Z at the end of the file then do `:setlocal noendofline nofixendofline`, you can write a file with ^Z.
For example, the following function might be used to append ^Z:

```vim
func AppendCtrlZ()
  let EOF = "\x1a"
  if search(EOF .. '\%$', 'n') == 0
    " Ctrl-Z not found.
    if &endofline
      call append(line('$'), EOF)
    else
      call setline('$', getline('$') .. EOF)
    endif
  endif
  setlocal noendofline nofixendofline
endfunc
```

**Comment by 0cjs** [comment]
I'm the one who originally asked about this problem [on the Vi and Vim StackExchange][38912]. To help understand the level of importance (or lack thereof :-)) of the ^Z EOF markers, here's what happens without it in various situations.

- CP/M absolutely requires this for text files as it measures file lengths in blocks, rather than bytes. That means that _every_ file is padded out to a multiple of 128 bytes or more, and a text file without a ^Z terminator will have garbage at the end when read.
- Newer disk operating systems (e.g., MS-DOS 1.0, released 1981) keep file length in bytes, so files without a terminator still have EOF at the proper location. Here the only issue is files that _do_ have a ^Z in them have EOF at the ^Z when read in text mode and EOF at the end of the file when read in binary mode. Files do exist designed for "dual-mode" use, with text at the start, followed by a ^Z, followed by binary data. Vim has no problem editing these because the ^Z, not being at the _end_ of the file, is not stripped off by the `filemode=dos` read.
- On Unix and modern systems, files with ^Z as an EOF marker are used mainly when the Unix/modern system is a host for cross-development with one of the above machines as a target. Here the issue is that you can end up with Git, etc. detecting and committing file changes when the files were not supposed to be changed. (This might be considered a relatively trivial problem when the target is an MS-DOS style system rather than CP/M.)


[38912]: https://vi.stackexchange.com/q/38912/15666

**Comment by 0cjs** [comment]
As far as a change/fix goes, here are my thoughts on it.

The _only_ issue here is the deletion of a final ^Z when reading a file in DOS mode. Once it's been read, the file is properly written if you have a ^Z at the end: simply re-adding the ^Z alone on the last line and executing `:setlocal noendofline | write` demonstrates this.

For those using `nofixeol` there's no reason (beyond aesthetic) to remove the ^Z, and some reason not to (see my comment above).

I suspect that the ^Z removal was added for those using `fixeol` (which is indeed the default mode) because writing out the file in that case adds _extra_ characters (the final CR, NL) to the end of the file. This won't change the file's interpretation for DOSes that use the ^Z EOF convention, but may change how it's interpreted on other systems, or for programs that implement their own, different text file reading code on top of the underlying DOS's binary read mechanism. (Arguably, the same could be true about the ^Z removal, however.)

The solution that seems best to me is not to remove the Ctrl-Z at EOF if it's there when the file is read, and to ensure that the file is written out (with or without ^Z at EOF) the same way as it was read in, unless the end of the file is modified. Simply not deleting the ^Z on read won't work in all situations, however; if the user has `fixeol` on (which is the default) _or_ has `eol` on (also the default), the end of the file will not be written out as it came in.

One can partly fix this by detecting the ^Z at the end of the file during read (though not deleting it) and then setting `nofixeol` and `noeol` so that if the user doesn't touch the ^Z it's written out properly. But the issue there is now that if the user decides to delete the ^Z she must also set `fixeol` to maintain the line terminator that would have been written out for what was the second-last line before the ^Z was deleted. This is not hard to do, but is probably not intuitive to many. Though at least there would be a `[noeol]` warning on file load.

One way of fixing the above would be to store a new value that indicates this situation, perhaps by expanding the values assignable to `eol` or perhaps by adding another variable. But of course nobody wants to add another variable if they don't have to.

How about changing the behaviour of `fixeol` in DOS mode? Rather than always restoring a missing EOL at EOF, it would instead look to see if the last character of the file is a ^Z and do the proper "DOS fix" based on that: always write without EOL at EOF if the last character is ^Z, and always write with EOL at EOF if the last character is anything else.

**Comment by k-takata** [comment]
Detecting a ^Z and restoring it before writing can be done by Vim script:

```vim
func CheckCtrlZ()
  let lastbyte = readblob(expand('%'))[-1]
  let b:has_ctrlz = lastbyte == 0x1a
endfunc

func AppendCtrlZ()
  let EOF = "\x1a"
  if search(EOF .. '\%$', 'n') == 0
    " Ctrl-Z not found.
    if &endofline
      call append(line('$'), EOF)
    else
      call setline('$', getline('$') .. EOF)
    endif
  endif
  setlocal noendofline nofixendofline
endfunc

augroup DOSFILE
  au!
  au BufReadPre * call CheckCtrlZ()
  au BufWritePre * if b:has_ctrlz | call AppendCtrlZ() | endif
augroup END
```

**Comment by 0cjs** [comment]
@k-takata Right, I'd not mentioned here (though I had [mentioned elsewhere](https://vi.stackexchange.com/a/38918/15666) that it seems one can use Vimscript to fix the problem. But your post raises the question: Is keeping the current behaviour of `filemode=dos` with regard to the Ctrl-Z EOF marker the correct thing to do? I'd argue that my approach above of changing the `fixeol` behaviour with `filemode=dos` makes Vim do the more correct thing, though I am not fully confident in that answer and would welcome discussion and corrections. (That said, I do welcome workarounds while we're waiting for a fix to Vim itself, if a change to Vim is the way we decide to go.)

With regard to your script itself, it has several failings. First, it applies to files even when they're not being in `filemode=dos`, and second it adds back a Ctrl-Z if one was originally there even if the user deleted it intentionally. These combine to actually break the final EOL in a DOS file if it was edited in `filemode=unix` to remove the Ctrl-Z. In the test below you'll note that the final EOL changed from CR LF to just CR.

```
$ echo; printf 'a\r\nb\r\n\x1a' > z; xxd -g1 z; vi z; xxd -g1 z
00000000: 61 0d 0a 62 0d 0a 1a                             a..b...
00000000: 61 0d 0a 62 0d 1a                                a..b..
```

Also, is that `readblob()` going to be efficient on large files? It looks as if it could be expensive.

**Comment by k-takata** [comment]
Changing the current behavior may break the compatibility. I'm not sure whether it is acceptable to fix this "rare" issue.

The above script is just an example to demonstrate that the issue can be solved with Vim script. It may have some corner cases, but they should be easily fixed, I think.
To remove ^Z, you can do `:let b:has_ctrlz=0` before writing the file.

> Also, is that readblob() going to be efficient on large files? It looks as if it could be expensive.

It might be better to add "offset" and "size" parameters to readblob().

**Comment by 0cjs** [comment]
@k-takata Well I am quite concerned about compatibility and I'm not advocating blindly making this change, but in case it wasn't clear from my posts above, _compatibility with CP/M files is already broken_ and has been since this behaviour was added in 1994.

I _think_ my proposed change does less non-user-requested changing of files than the current system, and the only compatibility it would break would be users who were expecting Vim to remove a character from a file _without them doing any editing commands_ to indicate they wanted it removed. That said, I'm totally open to corrections and discussion about the advantages and disadvantages of my proposal, other proposals and the current way of doing things.

I am getting the feeling you've already decided that the current behaviour should not be changed and feel fairly firmly about that. Is that the case, or are you open to changing it? If so, what are your concerns about changing it?

**Comment by chrisbra** [comment]
This is just my opinion, but the fact that this behaviour after 28 years now comes up, already indicates, that this is some kind of a rare niche problem, isn't it? 

If we change the current behaviour now, this would be a backwards incompatible change. So why do you want to keep the CTRL-Z? Is it you just need to use those files on CP/M? I am surprised this is actually used today :) 

Anyhow, I would rather suggest to either use `writefile()` using the `b` flag and then you can terminate your lines with CRLF and add a CTRL-Z at the end of file. Or perhaps post-processing the file using a perl or python script (I actually wonder how `unix2dos(1)` handles this case, it doesn't seem to have an option for that)


> users who were expecting Vim to remove a character from a file without them doing any editing commands to indicate they wanted it removed. 

You mean, whenever users issue a `:w` command? Or is vim really writing a file when not giving a *write* command?

**Comment by arp242** [comment]
> This is just my opinion, but the fact that this behaviour after 28 years now comes up, already indicates, that this is some kind of a rare niche problem, isn't it?
>
> If we change the current behaviour now, this would be a backwards incompatible change. So why do you want to keep the CTRL-Z? Is it you just need to use those files on CP/M? I am surprised this is actually used today :)

It's my impression there's a bit of renewed in interest in these systems; I think it's because something that's 10 or 15 years old is just "old", but if something is 30 years old it starts becoming "history". This is also why people have spent quite a bit of effort to recover early Unix versions, or why archaic scripts like Cuneiform are Unicode. No one uses Cuneiform to write emails to their friends, but having support for such old and archaic systems enables examining history (even the undeciphered Linear-A is in Unicode!)

Still a very rare use case of course, but I think that's why it came up now, and not 20 years ago.

**Comment by brammool** [comment]
> I'm the one who originally asked about this problem [on the Vi and Vim
> StackExchange][38912]. To help understand the level of importance (or
> lack thereof :-)) of the ^Z EOF markers, here's what happens without
> it in various situations.
> 
> - CP/M absolutely requires this for text files as it measures file
> lengths in blocks, rather than bytes. That means that _every_ file is
> padded out to a multiple of 128 bytes or more, and a text file without
> a ^Z terminator will have garbage at the end when read.

CP/M only provies about 56 Kbyte to the program.  Vim can't fit in there
(it stopped fitting in the 600 Kbyte that MS-DOS offers at some point).
Thus this is only for editing files that are written to floppy disk and
then used on a CP/M system (although I would assume nearly everybody
uses an emulator).

> - Newer disk operating systems (e.g., MS-DOS 1.0, released 1981) keep
> file length in bytes, so files without a terminator still have EOF at
> the proper location. Here the only issue is files that _do_ have a ^Z
> in them have EOF at the ^Z when read in text mode and EOF at the end
> of the file when read in binary mode. Files do exist designed for
> "dual-mode" use, with text at the start, followed by a ^Z, followed by
> binary data. Vim has no problem editing these because the ^Z, not
> being at the _end_ of the file, is not stripped off by the
> `filemode=dos` read.

If I recall correctly, when I was still using these sytems, keeping the
CTRL-Z at the end was totally optional.  In fact, it's better to drop it
since there are many more programs that have a problem with it than
programs that require it.  This is especially when taking a file from an
old system with the intention of using it on a newer system, which is
the most common.

> - On Unix and modern systems, files with ^Z as an EOF marker are used
> mainly when the Unix/modern system is a host for cross-development
> with one of the above machines as a target. Here the issue is that you
> can end up with Git, etc. detecting and committing file changes when
> the files were not supposed to be changed. (This might be considered a
> relatively trivial problem when the target is an MS-DOS style system
> rather than CP/M.)

I think for those few people who like to use these old systems, it's OK
to do some extra work.  So long as it's possible to do, with
autocommands, plugin, etc.  I do not see a reason to make it very easy.

-- 
Why is "abbreviation" such a long word?

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by k-takata** [comment]
> > Also, is that readblob() going to be efficient on large files? It looks as if it could be expensive.
> 
> It might be better to add "offset" and "size" parameters to readblob().

Added in v9.0.0795.
`CheckCtrlZ` can be updated like this:
```
func CheckCtrlZ()
  let b:has_ctrlz = readblob(expand('%'), -1) == 0z1a
endfunc
```

Someone could turn my script above into a complete plugin.
I don't see a reason to support this on Vim natively.

**Comment by brammool** [comment]
> I'd be interesting in hearing about what programs have to have the
> Ctrl-Z removed when present. The typical use case would seem to me to
> be taking an original file from a system that has a Ctrl-Z at EOF,
> editing it, and then using the edited version on the original system.
> If it doesn't work on the original system without removing the Ctrl-Z,
> the original file wouldn't have worked on that system, which doesn't
> make sense.

For a while old files from CP/M and DOS systems were copied to newer
systems and there the trailing CTRL-Z caused error messages.  This was
(and is) the most common situation, therefore the removal of the
trailing CTRL-Z.

> Do we all agree that we're not talking about adding a Ctrl-Z to new
> files, but merely preserving it in files that already have it unless
> the user explicitly deletes it?

I don't think that after all these years there now is a reason to change
the behavior.

-- 
MICHAEL PALIN PLAYED: 1ST SOLDIER WITH A KEEN INTEREST IN BIRDS, DENNIS, MR
                      DUCK (A VILLAGE CARPENTER WHO IS ALMOST KEENER THAN
                      ANYONE ELSE TO BURN WITCHES), THREE-HEADED KNIGHT, SIR
                      GALAHAD, KING OF SWAMP CASTLE, BROTHER MAYNARD'S ROOMATE
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shougo** [comment]
@0cjs @arp242 I have created the PR.  It is the desired behavior?
Please test.

**Comment by 0cjs** [comment]
@Shougo #11408 is not really what I was thinking of. It's still sometimes removing the Ctrl-Z in `readfile()`, which I don't find to be very straightforward behaviour because once you've read a file (in this case just with `fixeol` set) you have no idea, without doing some sort of external check, whether a Ctrl-Z terminates the file you read. Essentially, you've lost information and need to go back and read the file again to get it back. (And if you don't do that right away, the information is lost permanently unless you have another copy of the file.)

Further, assuming that text files both with and without a Ctrl-Z terminator are "proper" DOS files (in the generic sense, not specific to how Vim defines it), which position I agree with, what you need to do to "fixeol" the file depends on whether it has the terminator or not. A file without a Ctrl-Z terminator should have a CR LF after the last line so as not to have a partial last line, but a file with a Ctrl-Z terminator should have CR LF on what Vim considers to be the _second_ last line, and not have a line terminator after the Ctrl-Z (unless the user decides explicitly to put one there, in which case it's no longer a Ctrl-Z-terminated file, at least for those filesystems that support exact file lengths).

So what I think should happen is that readfile() _never_ removes any characters from the file (thus avoiding the data loss), and with `fileformat=dos` we want the following:
- `binary`: the current behaviour is correct; `fileformat` is ignored
- `nobinary`, `fixeol`:
  - if file ends with Ctrl-Z, write out the Ctrl-Z as the last character of the file, with no CR LF after it
  - if file does not end with Ctrl-Z, write out a CR LF at the end of the last line of the file
- `nobinary`, `nofixeol`: Use the current `eol` setting.

Note that the setting or resetting of `eol` on read is fine just as it is: it will always be off for a Ctrl-Z-terminated file, and a file containing Ctrl-Z CR LF at the end is not a Ctrl-Z terminated file but a non-Ctrl-Z-terminated file that happens to have a last line containing a Ctrl-Z.

This still has one case where the behaviour might be considered unintuitive. A Ctrl-Z-terminated file is read in `filemode=dos`, the setting becomes `noeol`, and the user then deletes the last line of the file with the Ctrl-Z terminator. At that point, if the user does not have `fixeol` set he gets a file missing the final CR LF that was in the original file (though one character before the end of the original file).

That's something I can live with, actually, though I wouldn't object to it being fixed. (This would probably require storing another setting.) My two main concerns are that a) nothing is deleted from the file without me explicitly requesting it be deleted, and b) with `fixeol` on reasonable things happen.

**Comment by 0cjs** [comment]
@Shougo Oh, after writing all that I see more clearly what I feel is a design inconsistency in your patch. `fixoel` is (currently, and as far as I know) a flag affecting _writes,_ and your patch is using it not at write time but at _read_ time. I think it's much simpler for the user to model what's going on if `fixeol` affects only writes. (And not following "always determine line/file endings on write, rather than read" is pretty much the core of this whole problem.)

**Comment by k-takata** [comment]
I've rewrote the script. Now it restore the ^Z after reading a file.

```vim
func RestoreCtrlZ()
  if &fileformat != 'dos' || readblob(expand('%'), -1) != 0z1a
    return
  endif

  let EOF = "\x1a"
  if search(EOF .. '\%$', 'n') == 0
    " Ctrl-Z not found.
    let save_mod = &modified
    if &endofline
      call append(line('$'), EOF)
    else
      call setline('$', getline('$') .. EOF)
    endif
    let &modified = save_mod 
  endif
  setlocal noendofline nofixendofline
endfunc

augroup DOSFILE
  au!
  au BufReadPost * call RestoreCtrlZ()
augroup END
```

**Comment by 0cjs** [comment]
@Shougo The load behaviour is as desired, yes. And with `nofixendofline` saving works fine too.

However, saving is broken with `fixendofline` when you leave the Ctrl-Z intact. The before and after contents of the file are:
```
00000000: 61 0d 0a 62 0d 0a 1a                             a..b...
00000000: 61 0d 0a 62 0d 0a 1a 0d 0a                       a..b.....
```
You'll notice it's added a `0d 0a` after the Ctrl-Z, meaning that this is no longer a Ctrl-Z-terminated file, but one where the last line contains Ctrl-Z. That's probably not acceptable behaviour to most people (though I could live with it).

So the other part that would be necessary here would be to change the fixendofline code so that if `filemode=dos` _and_ the last character in the buffer is a Ctrl-Z, it does not add a CR LF after that. (Otherwise it adds the CR LF to the end as usual.)

That still leaves people configured with `nofixeol` having to manually change `eol` to `noeol` if they remove a Ctrl-Z at the end, or `noeol` to `eol` if they add one. As a `nofixeol` user myself I personally don't have any issues with that, and others might agree that someone who sets `nofixeol` should know what he's doing. It is a pretty rare situation, and I would suspect that people who are changing the file ending would generally know what they're doing.

**Comment by Shougo** [comment]
My fix intension means just refactor the old code for `CP/M` than perfect support for `CP/M` emulator.

>  That's probably not acceptable behaviour to most people (though I could live with it).

It is acceptable.  Because it is noendofline file.  It is UNIX culture though.
It is like `CTRL-Z` terminate file behavior in `CP/M`.

> So the other part that would be necessary here would be to change the fixendofline code so that if filemode=dos and the last character in the buffer is a Ctrl-Z, it does not add a CR LF after that. (Otherwise it adds the CR LF to the end as usual.)

I can change the `fixeol` behavior, but it adds some complexity and it does not have many advantages for other users.

Users should set `nofixendofline` option if needed.
Vim users should know what do you do.

**Comment by 0cjs** [comment]
@Shougo So the main problem I was focusing on is that with the default Vim configuration, someone loads a CP/M text file that worked in their emulator, edits a single line of text (or even changes nothing at all) and then saves it will find that the edited file is broken for programs running under CP/M. Your change does fix that, but it then (again in the default config) will break behaviour in certain other situations where the intent is not to use the file on its original system but to convert it for use on a modern system.

You're correct that proper setting of options when you want to do this will solve the CP/M problem and, as I said, I'm ok with that. but I'm not sure that others are. (Perhaps someone who has seen situations where not removing a Ctrl-Z terminator breaks things could weigh in with an example, so we can better understand what's breaking there and how?)

I think that my suggested fixeol change has sufficient advantage to be worth implementing because it means that under default settings things continue to work in the most obvious way: Vim does the right thing with people deleting a Ctrl-Z EOF marker without any special work or knowledge on their part. And I'd actually be happy to implement this myself if there's agreement that it's likely to be an acceptable change. (Right now it looks to me as if it would be outright rejected.)

So where are we now with this? Is Shougo's current patch something that:
- could go in as-is now?
- could go in with a `fixeol` change to maintain current behaviour when the trailing Ctrl-Z is deleted in `filemode=dos`?
- will never be accepted?

**Comment by Shougo** [comment]
@0cjs Please see https://github.com/vim/vim/pull/11408#issuecomment-1287065133

**Comment by Shougo** [comment]
I don't have the big motivation for it.  But @0cjs can implement the feature.

**Comment by 0cjs** [comment]
> @0cjs Please see [#11408 (comment)](https://github.com/vim/vim/pull/11408#issuecomment-1287065133)

@Shougo Thanks for pointing me at this. I've [replied there](https://github.com/vim/vim/pull/11408#issuecomment-1293467453).

**Comment by k-takata** [comment]
Fixed by 9.0.0816, 9.0.0819 and 9.0.0826. Closing.
